app/src/main/java/com/keylesspalace/tusky/util/LinkHelper.kt

```diff
@@ -0,0 +1,264 @@
+                builder.getSpanEnd(span), builder.getSpanFlags(span))
+    }
+
+    val otherSpans = builder.getSpans(0, content.length, ForegroundColorSpan::class.java)
+    val usedMentionIds = HashSet<String>()
+
+    for (span in otherSpans) {
+        val start = builder.getSpanStart(span)
+        val end = builder.getSpanEnd(span)
+        val flags = builder.getSpanFlags(span)
+        val text = builder.subSequence(start, end)
+
+        val customSpan = when (text[0]) {
+            '#' -> getTagSpan(text.substring(1), listener)
+            '@' -> {
+                if (mentions != null && mentions.isNotEmpty()) {
+                    /* There may be multiple matches for users on different instances with the same
+                     * username. If a match has the same domain we know it's for sure the same, but if
+                     * that can't be found then just go with whichever one matched last. */
+                    val id: String? = firstUnusedMention(text.substring(1), mentions, usedMentionIds)
+                    if (id != null) {
+                        usedMentionIds.add(id)
+                        getAccountSpan(id, listener)
+                    } else {
+                        null
+                    }
+                } else {
+                    null
+                }
+            }
+            else -> null
+        }
+
+        replaceSpan(builder, span, customSpan, start, end, flags)
+    }
+
+    view.text = builder
+    view.linksClickable = true
+    view.movementMethod = LinkMovementMethod.getInstance()
+}
+
+/**
+ * Replace a span within a spannable string builder
+ * @param builder the builder to replace spans within
+ * @param oldSpan the span to be replaced
+ * @param newSpan the new span to be used
+ * @param start the start of the new span
+ * @param end the end of the new span
+ * @param flags the flags to be used for the new span
+ */
+private fun replaceSpan(builder: SpannableStringBuilder, oldSpan: CharacterStyle?, newSpan: ClickableSpan?, start: Int, end: Int, flags: Int) {
+    builder.removeSpan(oldSpan)
+    builder.setSpan(newSpan, start, end, flags)
+
+    /* Add zero-width space after links in end of line to fix its too large hitbox.
+     * See also : https://github.com/tuskyapp/Tusky/issues/846
+     *            https://github.com/tuskyapp/Tusky/pull/916 */
+    if (end >= builder.length || builder[end] == '\n') {
+        builder.insert(end, "\u200B")
+    }
+}
+
+/**
+ * Returns the first account id with matching username from mentions that isn't contained in usedIds,
+ * or the id of the last matching account, if all matching ids are already contained
+ * @param username the username to match
+ * @param mentions the mentions to search
+ * @param usedIds the collection of ids already used
+ */
+private fun firstUnusedMention(username: String, mentions: Array<Status.Mention>, usedIds: Collection<String>): String? {
+    var id: String? = null
```

Comment:
First or all, could be rewritten more idiomatic:

```
return mentions.firstOrNull { mention -> 
        mention.localUsername == username && usedIds.contains(mention.id)
    }?.id
```

Second, maybe it makes sense to:
1. In advance copy mentions list
2. Remove from it when we used a mention?

Maybe I misunderstand the algorithm but should be a lil cheaper and straightforward?

Reply:
The benefit is minimal and we should aim for readability here rather than performance I think so if my suggestion is clumsy feel free to ignore it